### ISSUE TITLE
Potential issue with getting offsets in the Fortran ceedvectorgetarray method

### ISSUE THREAD

**Issue by dmed256** [CEED/libCEED#488]
## Description

The current implementation of the `ceedvectorgetarray` method is: (I renamed `b` to `vecArray` in this code snippet to make the issue clearer)

```c
#define fCeedVectorGetArray FORTRAN_NAME(ceedvectorgetarray,CEEDVECTORGETARRAY)
void fCeedVectorGetArray(int *vec, int *memtype, CeedScalar *array,
                         int64_t *offset,
                         int *err) {
  CeedScalar *vecArray;
  CeedVector vec_ = CeedVector_dict[*vec];
  *err = CeedVectorGetArray(vec_, *memtype, &vecArray);
  *offset = vecArray - array;
}
```

There is an example in `t104-vector-f.f90` that uses it:

```f90
      call ceedvectorsetarray(x,ceed_mem_host,ceed_use_pointer,a,aoffset,err)
      call ceedvectorgetarray(x,ceed_mem_host,b,boffset,err)
```

I'm seeing this case fail only in the fortran test (maybe related to #184) and I don't understand the relationship between `array` and `vecArray`. 

In the example above, it looks like it's comparing two unrelated pointers inside `CeedVectorGetArray`: 
- `a` (`vecArray`) 
- `b` (`array`)

**Comment by jedbrown** [comment]
The issue here is that F77 has no concept of a pointer that can be changed, so instead we just compute the distance between `&a(1)` (the location of the actual memory) and `&b(1)` (a completely arbitrary place that we never access). The declaration can actually be
```fortran
      real*8 a(10)
      real*8 b(1)
```

We then do a grotesquely out-of-bound access to the memory in `a` via the pointer/array `b`:
```fortran
      ! b is input-only to ceedvectorgetarray; boffset is an output
      call ceedvectorgetarray(x,ceed_mem_host,b,boffset,err)
      b(boffset+3) = -3.14
```

This can be much cleaner if we used Fortran90, but the Nek team was adamant that Fortran90 is too new...

**Comment by dmed256** [comment]
Ah that makes a lot of sense, thanks!

**Comment by knepley** [comment]
If you ask me, F77 is a new fangled monstrosity strapped to the top of the industry standard Fortran 66.

**Comment by dmed256** [comment]
Ah, so I'm expecting `b` to change in the OCCA backend. So when this gets called
```f90
call ceedvectorrestorearray(x,b,boffset,err)
```
we're copying `b` to the vector `x`. In this case, we `b` is not really `a` so we're overriding the set `-3.14` value back to `0` (or w/e the Fortran `real` default is)

@jedbrown Should we instead change the implementation from
```c
void fCeedVectorRestoreArray(int *vec, CeedScalar *array,
                             int64_t *offset, int *err) {
  *err = CeedVectorRestoreArray(CeedVector_dict[*vec], &array);
  *offset = 0;
}
```
->

```c
void fCeedVectorRestoreArray(int *vec, CeedScalar *array,
                             int64_t *offset, int *err) {
  CeedScalar offsetArray = array + *offset;
  *err = CeedVectorRestoreArray(CeedVector_dict[*vec], &offsetArray);
  *offset = 0;
}
```

**Comment by jedbrown** [comment]
That suggested change seems reasonable to me.

**Comment by dmed256** [comment]
Cool, I added it to the PR in an individual commit

**Comment by jedbrown** [comment]
I'll close this because I think it has been resolved (and the small improvement is coming in #305, to be included in v0.6.1). Please reopen if needed.
